Library/Homebrew/extend/os/mac/keg_relocate.rb

```diff
@@ -45,27 +45,40 @@ def fix_dynamic_linkage
         change_dylib_id(dylib_id_for(file), file) if file.dylib?
 
         each_linkage_for(file, :dynamically_linked_libraries) do |bad_name|
-          # Don't fix absolute paths unless they are rooted in the build directory
-          next if bad_name.start_with?("/") &&
-                  !rooted_in_build_directory?(bad_name)
-
-          new_name = fixed_name(file, bad_name)
-          change_install_name(bad_name, new_name, file) if new_name != bad_name
+          # Don't fix absolute paths unless they are rooted in the build directory.
+          new_name = if bad_name.start_with?("/") && !rooted_in_build_directory?(bad_name)
+            bad_name
+          else
+            fixed_name(file, bad_name)
+          end
+          loader_name = loader_name_for(file, new_name)
+          change_install_name(bad_name, loader_name, file) if loader_name != bad_name
         end
 
         each_linkage_for(file, :rpaths) do |bad_name|
           # Strip duplicate rpaths and rpaths rooted in the build directory.
-          next if !rooted_in_build_directory?(bad_name) &&
-                  (file.rpaths.count(bad_name) == 1)
-
-          delete_rpath(bad_name, file)
+          if rooted_in_build_directory?(bad_name) || (file.rpaths.count(bad_name) > 1)
+            delete_rpath(bad_name, file)
+          else
+            loader_name = loader_name_for(file, bad_name)
+            change_rpath(bad_name, loader_name, file) if loader_name != bad_name
+          end
         end
       end
     end
 
     generic_fix_dynamic_linkage
   end
 
+  def loader_name_for(file, target)
+    # Use @loader_path-relative install names for other Homebrew-installed binaries.
+    if target.start_with?(HOMEBREW_PREFIX)
+      "@loader_path/#{Pathname(target).relative_path_from(file.dirname)}"
+    else
+      target
+    end
```

Comment:
I wonder if we could/should gate this behind an environment variable (e.g. `HOMEBREW_USE_RELATIVE_INSTALL_NAMES`) first?

We can set it in Homebrew/core CI for testing, and then remove the environment variable check when we're ready. But at least the env check will allow us to quickly turn it off if we discover any serious problems with it.

It'll also help us fine-tune the timing of when this change will take effect in Homebrew/core CI. I suspect we'll actually want to make sure that the `linkage_checker` changes lands in a release tag first before building bottles with this enabled.

Reply:
An alternative I can think of now (could be combined with an env var) that would cause this PR to take effect gradually would be to add an extra check in the code so that these new lines quoted here aren't run when pouring from a bottle (as of now, the code in the quoted snippet is called from `#fix_dynamic_linkage`, which can get called regardless of if installing from source or a bottle).

Would need some more changes in the code to achieve this (guess the simplest way would be to add a flag to `#fix_dynamic_linkage` that tells it if it's pouring from a bottle or not), but it would make it so that only newly built bottles and installs from source are affected by the change, containing the extent of immediate consequences of any negative impact from this PR. Also, as discussed above, bottles built after this change will already have relative install names in their binaries and don't really need to go through this rewrite step at all.

Reply:
> I wonder if we could/should gate this behind an environment variable (e.g. `HOMEBREW_USE_RELATIVE_INSTALL_NAMES`) first?

Agreed. This is a blocker for me, I think, sorry!

Ideally all the same behaviour would remain unless one or more environment variables are set. I am fine with there being one variable to handle bottle creation and another for bottle pouring or one variable for both; whatever consensus @carlocab @gerlero agree with works for me.

Reply:
@MikeMcQuaid Your call. I'm not sure then whether to (a) gate relative install names when pouring from bottles with the same env variable, (b) control what to do with bottle installs with a different environment variable, or (c) not do anything new at all when pouring from bottles.

(b) would be the most flexible but I don't that flexibility is really needed... (c) is the ultimate goal of this PR; but if the idea of introducing this behind an env var is for the feature to be tested more thoroughly, then (a) should create more opportunities for that.

@carlocab do you have a particular stance on this?

Reply:
@gerlero My take is that any functionality that changes with creating or pouring bottles should be behind 1 or 2 environment variables for now. I don't have any preference whether it's 1 or 2: you decide. If @carlocab replies today with a preference: do what he says otherwise you just do whatever you want (which sounds like 1 which is fine/good).

Thanks again, sorry for all the back and forth!

Reply:
> Suggestion: make `#loader_name_for` be no-op (i.e. always take the `else` branch) unless `HOMEBREW_RELOCATABLE_INSTALL_NAMES` is set.

Great. I can do that.

> If my understanding of how this all works is correct: the changes here should only matter when creating bottles, and not when pouring them, so there is no need for an extra environment variable that controls behaviour when pouring bottles.

Actually, the changes here as of now cover both, as `#fix_dynamic_linkage` is called when pouring bottles too:

https://github.com/Homebrew/brew/blob/a34d46e0938e2a830b8466729ffb2fb60de9a80e/Library/Homebrew/formula_installer.rb#L786

I apologize for this confusion; I got this part wrong when I first opened the PR (incorrectly assumed that `#fix_dynamic_linkage` was only invoked for builds from source), so this might be the source of the confusion.

As it is now, **bottles built without relative names will be changed to use relative names on install**. If this is undesired, then the new logic will need to be tweaked a bit to be able to detect the type of install and disable the behavior it it's a bottle pour.

> The `linkage_checker` changes are an improvement regardless of whether we adopt your `keg_relocate` changes or not, so these should be shipped to users as soon as it's sensible to.

I agree (that part could've even been its own PR). I'll move the new logic to `MachOShim` and then we should be done with that for now.

> When this happens, we can try setting `HOMEBREW_RELOCATABLE_INSTALL_NAMES` in Homebrew/core CI and see try these changes out. Once we're happy, we can remove the environment variable check.

Sounds good.

I'll let you know then I'm done with the new changes.

Reply:
> Actually, the changes here as of now cover both, as `#fix_dynamic_linkage` is called when pouring bottles too:

Yes, I noticed that too. My point is that after bottles are built with these changes: `fix_dynamic_linkage` should have no effect when pouring bottles.

Reply:
@MikeMcQuaid @carlocab Thanks again for all your comments. I've pushed the changes (added the `HOMEBREW_RELOCATABLE_INSTALL_NAMES` env var check and moved the expansion of relative install names to `MachOShim`).